demoapps/passwordlogin/java/org/openyolo/demoapp/passwordlogin/LoginViewModel.java

```diff
@@ -0,0 +1,513 @@
+/*
+ * Copyright 2017 The OpenYOLO Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.openyolo.demoapp.passwordlogin;
+
+import android.annotation.SuppressLint;
+import android.app.Application;
+import android.content.Intent;
+import android.databinding.ObservableBoolean;
+import android.databinding.ObservableField;
+import android.net.Uri;
+import android.support.annotation.AnyThread;
+import android.support.annotation.Keep;
+import android.support.annotation.MainThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.StringRes;
+import android.support.annotation.WorkerThread;
+import android.util.Log;
+import android.util.Patterns;
+import android.view.View;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.openyolo.api.CredentialClient;
+import org.openyolo.demoapp.passwordlogin.userdata.HashUtil;
+import org.openyolo.demoapp.passwordlogin.userdata.UserDataSource;
+import org.openyolo.protocol.AuthenticationDomain;
+import org.openyolo.protocol.AuthenticationMethods;
+import org.openyolo.protocol.Credential;
+import org.openyolo.protocol.CredentialRetrieveRequest;
+import org.openyolo.protocol.CredentialRetrieveResult;
+import org.openyolo.protocol.CredentialSaveRequest;
+import org.openyolo.protocol.CredentialSaveResult;
+import org.openyolo.protocol.Hint;
+import org.openyolo.protocol.HintRetrieveRequest;
+import org.openyolo.protocol.HintRetrieveResult;
+import org.openyolo.protocol.RetrieveBbqResponse;
+
+/**
+ * The view model for {@link LoginActivity}. Defines all the logic for authenticating a user
+ * which an email address and password, including interacting with any available OpenYOLO provider
+ * to directly retrieve or generate credentials.
+ */
+@WorkerThread
+public final class LoginViewModel extends ObservableViewModel {
+
+    private static final String TAG = "LoginViewModel";
+
+    private static final int MIN_PASSWORD_LENGTH = 4;
+
+    /**
+     * Indicates whether a progress bar should be displayed to the user, while an asynchronous
+     * request is made.
+     */
+    public final ObservableBoolean loading = new ObservableBoolean(true);
```

Comment:
Nit: rename isLoading

Reply:
Compromise: renamed to showLoading (to make it clear that this impacts the UI). Generally I prefer to reserve "is" prefixes for boolean getter methods.